core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -350,6 +350,16 @@ public static SqlNode convertConditionToSqlNode(RexNode node,
    */
   private static RexNode stripCastFromString(RexNode node, SqlDialect dialect) {
     switch (node.getKind()) {
+    case IS_NULL:
```

Comment:
the problem is that an expression such as 10 = '10' can be interpreted in two ways:
`CAST(10 AS VARCHAR) = '10'` or `10 = CAST('10' AS INT)`
To strip the cast we have to know what the target dialect would do.
These will NOT produce the same result for all expressions.
In particular, CAST to INT will produce a runtime exception if the value is not a number, while CAST to VARCHAR will never produce a runtime exception.

Reply:
Yes, I agree with it, but some dialect had restricted conversion conditions, such as Presto which not allowed varchar type convert to numberic type in supportsImplicitTypeCoercion method(maybe is aim to resolve the problem you mentioned), 
so do you suggest to do more cast handle for the problem？ @mihaibudiu

Reply:
Are you saying that in Presto there is no explicit cast from VARCHAR to a numeric type? 

I think that only the dialect can decide which cast to strip.
I don't think this logic can be in SqlImplementor.

Reply:
Yes, Presto had restricted cast the type from string to numberic, and this logic should not do this, maybe I need a more deepthinking about it, thanks for guidance! @mihaibudiu